app/controllers/openid_application_controller.rb

```diff
@@ -35,8 +36,11 @@ def authenticate_token
     return false if token.blank?
 
     # issued for a client vs a user
-    if token.client_credentials_token?
+    if token.client_credentials_token? || ssoi_token?
       token.payload[:icn] = fetch_smart_launch_context if token.payload['scp'].include?('launch/patient')
+      if token.payload['scp'].include?('launch')
+        token.payload[:launch] = JSON.parse(Base64.decode64(fetch_smart_launch_context))
```

Comment:
Because we don't _enforce_ that launch must be b64, what is the behavior here if it isn't?

Reply:
Good catch.  The JSON parse would throw an exception.

I have added a check to first check if valid JSON, otherwise it is assumed to be a plaintext ICN.